internal/commands/build/build.go

```diff
@@ -0,0 +1,233 @@
+package build
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/deislabs/cnab-go/bundle"
+	cnab "github.com/deislabs/cnab-go/driver"
+	"github.com/docker/app/internal/packager"
+	"github.com/docker/app/types"
+	"github.com/docker/buildx/build"
+	"github.com/docker/buildx/driver"
+	_ "github.com/docker/buildx/driver/docker" // required to get default driver registered, see driver/docker/factory.go:14
+	"github.com/docker/buildx/util/progress"
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/distribution/reference"
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/session/auth/authprovider"
+	"github.com/moby/buildkit/util/appcontext"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+type buildOptions struct {
+	noCache  bool
+	progress string
+	pull     bool
+	tag      string
+	out      string
+}
+
+func Cmd(dockerCli command.Cli) *cobra.Command {
+	var opts buildOptions
+	cmd := &cobra.Command{
+		Use:     "build [APPLICATION]",
+		Short:   "Build service images for the application",
+		Example: `$ docker app build myapp.dockerapp`,
+		Args:    cli.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			tag, err := runBuild(dockerCli, args[0], opts)
+			if err == nil {
+				fmt.Printf("Successfully build %s\n", tag.String())
+			}
+			return err
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.BoolVar(&opts.noCache, "no-cache", false, "Do not use cache when building the image")
+	flags.StringVar(&opts.progress, "progress", "auto", "Set type of progress output (auto, plain, tty). Use plain to show container output")
+	flags.BoolVar(&opts.pull, "pull", false, "Always attempt to pull a newer version of the image")
+	flags.StringVarP(&opts.out, "output", "o", "", "Dump generated bundle into a file")
+	flags.StringVarP(&opts.tag, "tag", "t", "", "Name and optionally a tag in the 'name:tag' format")
+
+	return cmd
+}
+
+func runBuild(dockerCli command.Cli, application string, opt buildOptions) (reference.Named, error) {
+	ctx := appcontext.Context()
+	info, err := dockerCli.Client().Info(ctx)
+	if err != nil {
+		return nil, err
+	}
+	majorVersion, err := strconv.Atoi(info.ServerVersion[:strings.IndexRune(info.ServerVersion, '.')])
+	if err != nil {
+		return nil, err
+	}
+	if majorVersion < 19 {
+		return nil, errors.New("'build' require docker engine 19.03 or later")
+	}
```

Comment:
I think this is something that should be extracted from the main function.
All we need to know when we are reading `runBuild` is we are checking if we run on the right version of the engine.

So by example something like:

```go
func runBuild(dockerCli command.Cli, application string, opt buildOptions) (reference.Named, error) {
        ctx := appcontext.Context()

        if err := checkAvailableDockerEngine; err != nil {
                return nil, err
        }

        // ...
}

func checkAvailableDockerEngine(ctx context.Context, dockerCli command.Cli) error {
	info, err := dockerCli.Client().Info(ctx)
	if err != nil {
		return err
	}
	majorVersion, err := strconv.Atoi(info.ServerVersion[:strings.IndexRune(info.ServerVersion, '.')])
	if err != nil {
		return err
	}
	if majorVersion < 19 {
		return errors.New("'build' require docker engine 19.03 or later")
	}

	return nil
}
```

Reply:
In a Teletubbies world, I would like the Plugin framework let us register commands with such version constraint, and report end-user "build command is disable. Require docker: 19.03 or later" so we don't have to bother with it :P